app/collections.py

```diff
@@ -1,97 +1,72 @@
-    def _get_generic_cls_args(cls):
-        return next(
-            base_cls.__args__
-            for base_cls in cls.__orig_bases__
-            if base_cls.__origin__ is BaseCollection
-        )
-
-    @property
-    def model_cls(self) -> type[ModelT]:
-        return self._get_generic_cls_args()[0]
-
-    @property
-    def model_create_cls(self) -> type[ModelCreateT]:  # pragma: no cover
-        return self._get_generic_cls_args()[1]
-
-    @property
-    def model_update_cls(self) -> type[ModelCreateT]:  # pragma: no cover
-        return self._get_generic_cls_args()[2]
+    def __init__(self, repository: Repository[ModelT]):
+        self.repository = repository
 
-    async def create(self, data: ModelCreateT) -> ModelT:
-        obj = self.model_cls(**data.model_dump())
-        self.session.add(obj)
-        await self.session.commit()
-        await self.session.refresh(obj)
-
-        return obj
+    def __init_subclass__(cls) -> None:
+        orig_bases = getattr(cls, "__orig_bases__", None)
+        if orig_bases is None:  # pragma: no cover
+            raise ValueError(f"Collection {cls.__name__} has no model class")
+        cls.model_cls = get_args(orig_bases[0])[0]
 
     async def get(self, id: str | UUID) -> ModelT:
-        obj = await self.session.get(self.model_cls, id)
-
-        if obj is None:
+        try:
+            return await self.repository.get(id)
+        except NotFoundError as e:
             raise HTTPException(
                 status_code=http_status.HTTP_404_NOT_FOUND,
-                detail=f"{self.model_cls.__name__} with ID {str(id)} wasn't found",
-            )
-
-        return obj
-
-    async def fetch_all(self) -> Sequence[ModelT]:
-        results = await self.session.exec(select(self.model_cls))
-        return results.all()
+                detail=str(e),
+            ) from e
 
     async def fetch_page(
         self, pagination_params: LimitOffsetParams | None = None
     ) -> LimitOffsetPage[ModelT]:
-        return await paginate(self.session, self.model_cls, pagination_params)
-
-    async def update(self, id: str | UUID, data: ModelUpdateT) -> ModelT:
-        statement = select(self.model_cls).where(self.model_cls.id == id)
-        results = await self.session.exec(statement)
-
-        obj: ModelT | None = results.first()
-
-        if obj is None:
-            raise HTTPException(
-                status_code=http_status.HTTP_404_NOT_FOUND,
-                detail=f"{self.model_cls.__name__} with ID {str(id)} wasn't found",
-            )
+        return await paginate(
+            self.repository.session,
+            self.repository.model_cls,
+            pagination_params,
+        )
 
-        for key, value in data.model_dump(exclude_unset=True).items():
-            setattr(obj, key, value)
 
-        self.session.add(obj)
-        await self.session.commit()
-        await self.session.refresh(obj)
+class CreateMixin[ModelT: UUIDModel, ModelCreateT: SQLModel]:
```

Comment:
Yeah mixins are probably the best approach here. Yesterday I was experimenting with the following as an alternative:

```python
def can_create(method):
    @functools.wraps(method)
    def inner(self: BaseCollection, *args, **kwargs):
        self.model_create_cls  # raises TypeError if not supported

        return method(self, *args, **kwargs)
    return inner

def can_update(method):
    @functools.wraps(method)
    def inner(self: BaseCollection, *args, **kwargs):
        self.model_update_cls  # raises TypeError if not supported

        return method(self, *args, **kwargs)
    return inner

class BaseCollection[ModelT: UUIDModel, ModelCreateT: SQLModel, ModelUpdateT: SQLModel]:  # same as before
    @property
    def model_cls(self) -> type[ModelT]:
        return self._get_generic_cls_args()[0]

    @property
    def model_create_cls(self) -> type[ModelCreateT]:  # pragma: no cover
        if len(self._get_generic_cls_args()) < 2:
            raise TypeError("Create operations are not supported for this collection")

        return self._get_generic_cls_args()[1]

    @property
    def model_update_cls(self) -> type[ModelCreateT]:  # pragma: no cover
        if len(self._get_generic_cls_args()) < 3:
            raise TypeError("Update operations are not supported for this collection")

        return self._get_generic_cls_args()[2]

    # the rest of the methods just like before
```

The only problem I could see was that `EntitlementCollection` would still have the `update` method (even if fails immediately on runtime).

Mixins solve that but they get more awkward to type:

```python
class EntitlementCollection(
    BaseCollection[Entitlement],
    CreateMixin[Entitlement, EntitlementCreate],
    UpdateMixin[Entitlement, EntitlementUpdate],
):
```

vs 

```python
class EntitlementCollection(BaseCollection[Entitlement, EntitlementCreate]):
    pass
```

No need to for a change, just sharing this and curious about your thoughts